nntrainer/dataset/batch_queue.cpp

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   batch_queue.cpp
+ * @date   13 July 2021
+ * @brief  This file contains thread safe queue
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <batch_queue.h>
+#include <chrono>
+
+#include <mutex>
+#include <nntrainer_error.h>
+#include <shared_mutex>
+
+using namespace std::literals::chrono_literals;
+
+namespace nntrainer {
+
+BatchQueue::BatchQueue(unsigned int queue_capacity_) :
+  queue_capacity(queue_capacity_) {
+  NNTR_THROW_IF(queue_capacity == 0, std::invalid_argument)
+    << "queue capacity of zero not supported!";
+}
+
+BatchQueue::BatchQueue(const BatchQueue &rhs) :
+  queue_capacity(rhs.queue_capacity) {}
+
+BatchQueue &BatchQueue::operator=(const BatchQueue &rhs) {
+  if (this == &rhs) {
+    return *this;
+  }
+  this->queue_capacity = rhs.queue_capacity;
+  return *this;
+}
+
+void BatchQueue::wait_and_push(T &&data) noexcept {
+  std::unique_lock<std::shared_mutex> lk(q_mutex);
+  q_writer_cv.wait(lk, [this] { return q.size() != queue_capacity; });
+  q.push(std::make_unique<T>(data));
+  q_reader_cv.notify_one();
+}
+
+std::unique_ptr<BatchQueue::T> BatchQueue::wait_and_pop() noexcept {
+  std::unique_lock<std::shared_mutex> lk(q_mutex);
+  q_reader_cv.wait(lk, [this] { return !q.empty(); });
+
```

Comment:
I think just checking if the queue is empyt is not enought. We need to check the number of remaining queue size here. If it is less than batch size, we need to wait longer.

Reply:
one data point here is a single batch so that, if q is not empty, it means we have a batch fully ready :)

Reply:
I see. Thanks.